### ISSUE TITLE
Stencil doesn't work in iOS Safari <= 10.3.1 (and possibly < 11)

### ISSUE THREAD

**Issue by cmaas** [ionic-team/stencil#1900]
**Stencil version:**

```
@stencil/core@1.5.2
@ionic/core@4.9.1
```

**I'm submitting a:**
[X] bug report
[ ] feature request
[ ] support request

**Current behavior:**
Stencil used to work in Safari 10.1+, but doesn't in the current version, because of a JS error. Older versions (Stencil 1.3.3 for example) still work.

**Expected behavior:**
Make it work again in iOS 10.

**Steps to reproduce:**
I created a new project with `ionic-pwa`, bumped the dependencies to the current version and uploaded the project here: https://stencil-ios01.netlify.com/
I installed a bunch of iOS Simulators to test it (iOS 9, 10.0, 10.1, 10.2, 10.3.1). Unfortunately, there are no simulators for the recently released iOS 10.3.4. If anyone has an older device (iPhone 5?) I'd appreciate a test on a device running iOS 10.3.4.

**Results:**
The "new" error is the first one (`TypeError: invalid scheme`). The other error (`Cannot declare a let variable twice` is a bug in Safari 10.3.1, but I'm not sure if it was fixed in 10.3.4: https://github.com/babel/minify/pull/567

iOS 10.3.1 – JS errors:
<img width="657" alt="10 3 1-error" src="https://user-images.githubusercontent.com/6094364/65752380-f3f09c00-e10c-11e9-890b-74f161d977fc.png">

Line of code referenced of first error:
<img width="472" alt="10 3 1-code" src="https://user-images.githubusercontent.com/6094364/65752389-f8b55000-e10c-11e9-80eb-12fe49d606bc.png">

iOS 10.0-10.2 (same first error as with 10.3.1):
<img width="500" alt="10 2-error" src="https://user-images.githubusercontent.com/6094364/65752334-df140880-e10c-11e9-9c07-941c13799f46.png">

iOS 9 – won't show because of CSS errors with variables.

Edit: I updated this issue a lot, after I tested with an older Stencil version, where everything works more or less as expected.

**Comment by cmaas** [comment]
**Update:** I installed an older version of Stencil (1.3.3). With this, it works at least somewhat on iOS 10.0. So the bug must've been introduced in a more recent version of Stencil. Unfortunately, Xcode 11.0 doesn't let you download iOS 10.2 or 10.1 anymore.

**Comment by manucorporat** [comment]
can you give  1.5.3 a try?

**Comment by cmaas** [comment]
@manucorporat Uploaded here: https://stencil-ios02.netlify.com
Now **silently** fails to load on iOS 10.0 and iOS 10.3.1 without any JS errors, but shows a blank page only.

Couldn't get to try it in iOS 10.1 and 10.2 so far, sorry.

**Comment by cmaas** [comment]
Fails silently on iOS 10.1 as well

**Comment by cmaas** [comment]
1.5.4 is still broken on iOS 10.1 – I updated the page here: https://stencil-ios02.netlify.com/

**Comment by viljamis** [comment]
Latest versions fails silently on 10.3.3 as well.

**Comment by cmaas** [comment]
@viljamis Oh great, someone with an actual 10.3.3 device. Glad you're here and can test 😄

**Comment by viljamis** [comment]
@cmaas Just tested on an actual device running 10.3.4 as well. Fails there too.

**Comment by cmaas** [comment]
1.7.0 still fails silently on iOS 10.1

**Comment by tricki** [comment]
Same for iOS 10.0.1, see https://github.com/ionic-team/stencil/issues/1963

**Comment by viljamis** [comment]
Ran out of time today while debugging this. But the issue is basically that Safari 10.x doesn’t support Stencil’s ESM build and just silently fails when trying to execute that. There seems to be multiple compatibility issues/features used that Safari 10.x does not support. Further, the “nomodule” build of Stencil has an IF statement in the beginning of it that prevents Safari 10.x from executing it. 

The non-ESM build works in Safari 10.x though (seems to work perfectly even in Safari 9.x?!) and if using script tags on a basic webpage there is a way around this issue. This is a bit hacky though but I tested it to be working on older Edge, IE11, Safari 9-13, Chrome, Firefox and Opera: 

```html
<!DOCTYPE html>
<meta charset="UTF-8" />
<head>
  <script>
    /**
     * Safari 10.xx supports <script type="module">, but since the module build just silently
     * fails, this code makes sure that we load the nomodule version in older Safari instead.
     *
     * The below if-statement detects other browsers, except Safari 10.x and under:
     *
     * - History.scrollRestoration was only introduced in Safari 11.0.
     * - "onbeforeload" is a feature only found from Safari.
     */
    if (history.scrollRestoration || !("onbeforeload" in document.createElement("script"))) {
      // Builds regular module and nomodule script tags for most browsers
      var m = document.createElement("script")
      m.src = "/build/duet.esm.js"
      m.type = "module"
      document.head.appendChild(m)
      var n = document.createElement("script")
      n.src = "/build/duet.js"
      n.setAttribute("nomodule", true)
      document.head.appendChild(n)
    } else {
      // For Safari 10.x build a basic script tag that loads the polyfilled version
      var s = document.createElement("script")
      s.src = "/build/duet.js"
      document.head.appendChild(s)
    }
  </script>
</head>
<body>
  <duet-button>Stencil.js component that now works in Safari 10.x</duet-button>
</body>
```

Wondering now though if @Manu or someone else in the Stencil team has ideas on how we could make Safari 10.x load the non-ESM build instead in Stencil itself. The IF statement I have in the above example should work well enough for this use case as history.scrollRestoration was only introduced in Safari 11.

**Comment by bbellmyers** [comment]
This is a major blocker for us upgrading from 1.1.1 to the latest Stencil version.  This is happening on IE11 Storybook/React as well.  https://stencil-worldwide.slack.com/archives/C79EANFL7/p1570827052090600

<img width="889" alt="Stencil-1 7 4-versus-IE11-Storybook" src="https://user-images.githubusercontent.com/9705513/68302383-90408380-0067-11ea-9889-02401af6ba96.png">

**Comment by danjohnson95** [comment]
Similar to @viljamis, we're currently working around this issue by only inserting the non-ESM build for Safari 10.x

```javascript
var isSafari10 = window.navigator.userAgent.indexOf('Safari') > 0 && window.navigator.userAgent.indexOf('Version/10') > 0;
var esmBuild = document.createElement('script');
var nonEsmBuild = document.createElement('script');

esmBuild.setAttribute('src', '/build/components.esm.js');
nonEsmBuild.setAttribute('src', '/build/components.js');

if (isSafari10) {
    document.head.appendChild(nonEsmBuild);
} else {
    esmBuild.setAttribute('type', 'module');
    nonEsmBuild.setAttribute('nomodule', true);

    document.head.appendChild(esmBuild);
    document.head.appendChild(nonEsmBuild);
}
```

**Comment by dutscher** [comment]
@CSchulz the problem is that is loaded but not executed due the "nomodule" attribute.

**Comment by dutscher** [comment]
How to implement the modules right isnt well documented from ionic.
See here: https://stenciljs.com/docs/javascript
And here: https://github.com/ionic-team/stencil#hosting-the-app
But here we can found: https://stenciljs.com/docs/output-targets#differential-bundling

And how they tested their Safari 10 bugfix? https://github.com/ionic-team/stencil/blob/master/CHANGELOG.md#-153-2019-09-27

cheers

**Comment by dutscher** [comment]
btw safari 10.0 on ios has a old webkit.performance api https://developer.mozilla.org/de/docs/Web/Web_Components

and breaks silently at `performance.mark()`, `performance.measure` & `performance.getEntriesByName`

to get rid of use `stencil build --no-profile`
thats the solution for old safari

cheers

**Comment by adamdbradley** [comment]
Thanks everyone for helping to debug this! @viljamis awesome job tracking down how to test for safari 10. The fix deals with:

1) Safari 10 supports ES modules
1) Safari 10 does not prevent downloading `nomodule` scripts (it'll request both esm and systemjs)
1) Safari 10 does not support `async/await`

So all that combined, it's best to always have Safari 10 use the es5/system build. The problem is that it'll always request both the esm and systemjs builds, so in this special case, we want the esm build to not continue to execute, while the systemjs should. Luckily this didn't add too much code, and when the day comes we do not have to worry about supporting Safari 10, these `if` statements can someday be removed. 

We'll be able to get this shipped in both the next `1.8.x` release and `1.9.x` prereleases.

Thanks everyone!

**Comment by cmaas** [comment]
Perfect, I can confirm that Stencil 1.8.5 works in iOS 10.3.1. Thanks for the work!
